core/js/src/main/scala/scalaz/zio/internal/impls/EnvJS.scala

```diff
@@ -0,0 +1,78 @@
+package scalaz.zio.internal.impls
+
+import java.util
+import java.util.concurrent.atomic.AtomicInteger
+
+import scalaz.zio.Exit.Cause
+import scalaz.zio.IO
+import scalaz.zio.duration.Duration
+import scalaz.zio.internal.Scheduler.CancelToken
+import scalaz.zio.internal.{ Env, Executor, Scheduler }
+
+import scala.concurrent.ExecutionContext.Implicits
+import scala.scalajs.js
+
+object EnvJS {
+
+  /**
+   * Creates a new default environment.
+   */
+  final def newDefaultEnv(reportFailure0: Cause[_] => IO[Nothing, _]): Env =
+    new Env {
+      val executor = Executor.fromExecutionContext(Executor.Yielding, 1024)(Implicits.global)
+
+      def executor(tpe: Executor.Role): Executor =
+        executor
+
+      val scheduler = newDefaultScheduler()
+
+      def nonFatal(t: Throwable): Boolean =
+        !t.isInstanceOf[VirtualMachineError]
+
+      def reportFailure(cause: Cause[_]): IO[Nothing, _] =
+        reportFailure0(cause)
+
+      def newWeakHashMap[A, B](): util.Map[A, B] =
+        new util.HashMap[A, B]()
```

Comment:
JS has a `WeakMap`. Might be worth using it if it's available at runtime and falling back to a normal HashMap otherwise.

Reply:
Unfortunately `WeakMap` and `WeakSet` aren't iterable, so cannot be used for supervisor: 
`supervisor: Iterable[Fiber[_, _]] => IO[Nothing, _]`...

Found this issue raised by @jdegoes - https://github.com/scala-js/scala-js/issues/3503

Reply:
Could this be usable somehow?

https://github.com/scala-native/scala-native/blob/cddb41c2ded4da43e0763d7ab65bdd3e49ad2d7e/javalib/src/main/scala/java/util/WeakHashMap.scala

Reply:
That just wraps a HashMap, it's not "weak" at all.

Reply:
I can delete this weak set entirely, it requires a bit of refactoring though. Definitely worth doing after this PR.